### PR TITLE
Enhance get_content() by omitting the version_membership lookup

### LINKED ISSUE #3969: Querying content in repository versions with lots of content is slow
**Version**
All

**Describe the bug**
When a repository version has upwards of 50k content in it, querying the content in the repository using `repo_version.get_content()` becomes extremely slow. In my testing the results haven't been consistent, however I have notice queries that can take anywhere from 200ms to 1000ms. This appears to be due to the complex joins that have to be made on version_added and version_removed.

**To Reproduce**
- Load up a repository with a lot of content
- Perform a query to fetch the last 100 objects or count how many objects are in it
- Observe query speeds that would make sonic the hedgehog faint

**Expected behavior**
Database queries should be fast. Hopefully less than 50ms for complex ones, ideally lower than that.

**Some possible solutions**
- Adding the repo version numbers directly to the RepositoryContent table speeds up queries greatly. My testing showed that this can lower the queryset speed to 80ms with 65k content in some cases.
- pulp_ansible is experimenting with querying the repository version table to look up RepositoryContent (https://github.com/pulp/pulp_ansible/blob/main/pulp_ansible/app/utils.py#L5). Results have been mixed with this. In some cases the query efficiency is substantially higher (on the order of 8x improvement). In other cases, it's roughly the same.
- Some form of memoization/caching of querysets.

### PR DESCRIPTION
**Pull request by lubosmj**
>>> timeit.timeit("len(v.get_content(Package.objects).only('pk'))", globals=globals(), number=10)
105.0813707895577
>>> timeit.timeit("len(v.get_content2(Package.objects).only('pk'))", globals=globals(), number=10)
58.41291802749038
>>> timeit.timeit("v.get_content(Package.objects).count()", globals=globals(), number=10)
62.557361379265785
>>> timeit.timeit("v.get_content2(Package.objects).count()", globals=globals(), number=10)
16.70007463544607

Credits to @newswangerd.

closes #3969

### CODE REVIEW

**Comment by lubosmj** [comment]
@ipanova and I experimented with the query on top of her publish task changes. The memory usage was not changed: https://github.com/pulp/pulp_rpm/pull/3224#issuecomment-1677252903. The same goes for the time spent in the execution.

**Comment by lubosmj** [comment]
I am going to compare the outputs of filters on the testing machine to be 100% sure there are no differences. This will include removing/adding content from a repository and filtering the content. A simple python script using the bindings might be sufficient.

**Comment by dralley** [comment]
It might be easier to just use the ORM directly, rather than bindings.  Otherwise you'll have a lot of pagination to deal with, and extra costs from all the serialization etc.  You only really need to know that the content sets are the same, which you can determine using the PKs alone.

**Comment by lubosmj** [comment]
@dralley, do you find these results satisfactory? `v.get_content2` stands for the updated version.

```
>>> v.get_content(Package.objects).order_by("pk").values_list("pk")
<BulkTouchQuerySet [(UUID('0000057e-3b84-4007-8a71-01aa312ed467'),), (UUID('000010d1-2f86-44d8-a310-8d0afc110985'),), (UUID('00001899-c33f-4889-b227-1e84479cb48f'),), (UUID('000020c7-8fc5-40e3-8058-cb9f0b6977f3'),), (UUID('00005bcf-a478-4d6e-a28c-92614b25b30a'),), (UUID('000070c5-1050-425d-b43e-7560cc679e44'),), (UUID('00007b2f-07a7-4568-8553-c55d58be06a7'),), (UUID('000086d8-fbb4-4874-a1ad-844ac3d27386'),), (UUID('00008e20-89de-416c-b09d-2581e4536866'),), (UUID('0000ea67-c0c8-43c8-bac2-4695e63ad43b'),), (UUID('0000fc5c-b24d-41ad-b8ce-01f7647501a2'),), (UUID('000171e0-635a-4fa2-94ff-baf3c39939ac'),), (UUID('000241e8-79ab-48db-a5dd-3153bd45f523'),), (UUID('00025b35-c2fb-47e4-b5da-bb3ebf4fd010'),), (UUID('000295a4-454a-4851-b8a3-c73abaef09e8'),), (UUID('00032ebb-fcaf-4a57-a760-d9b570b20058'),), (UUID('0003440c-cf2e-4f26-8ded-0f545321ee80'),), (UUID('00037c96-d149-4b32-9f0e-918d9983b7df'),), (UUID('000398d3-761c-4a0e-adb5-1a2111f3e1bd'),), (UUID('00039a33-30fd-4234-9e4d-d9805a14825e'),), '...(remaining elements truncated)...']>
>>> v.get_content2(Package.objects).order_by("pk").values_list("pk")
<BulkTouchQuerySet [(UUID('0000057e-3b84-4007-8a71-01aa312ed467'),), (UUID('000010d1-2f86-44d8-a310-8d0afc110985'),), (UUID('00001899-c33f-4889-b227-1e84479cb48f'),), (UUID('000020c7-8fc5-40e3-8058-cb9f0b6977f3'),), (UUID('00005bcf-a478-4d6e-a28c-92614b25b30a'),), (UUID('000070c5-1050-425d-b43e-7560cc679e44'),), (UUID('00007b2f-07a7-4568-8553-c55d58be06a7'),), (UUID('000086d8-fbb4-4874-a1ad-844ac3d27386'),), (UUID('00008e20-89de-416c-b09d-2581e4536866'),), (UUID('0000ea67-c0c8-43c8-bac2-4695e63ad43b'),), (UUID('0000fc5c-b24d-41ad-b8ce-01f7647501a2'),), (UUID('000171e0-635a-4fa2-94ff-baf3c39939ac'),), (UUID('000241e8-79ab-48db-a5dd-3153bd45f523'),), (UUID('00025b35-c2fb-47e4-b5da-bb3ebf4fd010'),), (UUID('000295a4-454a-4851-b8a3-c73abaef09e8'),), (UUID('00032ebb-fcaf-4a57-a760-d9b570b20058'),), (UUID('0003440c-cf2e-4f26-8ded-0f545321ee80'),), (UUID('00037c96-d149-4b32-9f0e-918d9983b7df'),), (UUID('000398d3-761c-4a0e-adb5-1a2111f3e1bd'),), (UUID('00039a33-30fd-4234-9e4d-d9805a14825e'),), '...(remaining elements truncated)...']>
>>> list(v.get_content2(Package.objects).order_by("pk").values_list("pk")) == list(v.get_content(Package.objects).order_by("pk").values_list("pk"))
True
>>> 
>>> list(v.get_content2(Package.objects).order_by("-pk").values_list("pk")) == list(v.get_content(Package.objects).order_by("-pk").values_list("pk"))
True
>>> with v.repository.new_version() as new_version:
...     new_version.remove_content(Package.objects.filter(pk="0000057e-3b84-4007-8a71-01aa312ed467"))
... 
>>> v.repository.latest_version()
<Repository: @rubygems/rubygems/fedora-rawhide-x86_64; Version: 2>
>>> v.repository.latest_version().get_content(Package.objects).filter(pk="0000057e-3b84-4007-8a71-01aa312ed467")
<BulkTouchQuerySet []>
>>> v.repository.latest_version().get_content2(Package.objects).filter(pk="0000057e-3b84-4007-8a71-01aa312ed467")
<BulkTouchQuerySet []>
>>> list(v.repository.latest_version().get_content2(Package.objects).order_by("-pk").values_list("pk")) == list(v.repository.latest_version().get_content(Package.objects).order_by("-pk").values_list("pk"))
True
>>> with v.repository.new_version() as new_version:
...     new_version.add_content(Package.objects.filter(pk="0000057e-3b84-4007-8a71-01aa312ed467"))
... 
>>> v.repository.latest_version().get_content(Package.objects).filter(pk="0000057e-3b84-4007-8a71-01aa312ed467")
<BulkTouchQuerySet [<Package (pulp_type=rpm.package): pk=0000057e-3b84-4007-8a71-01aa312ed467>]>
>>> v.repository.latest_version().get_content2(Package.objects).filter(pk="0000057e-3b84-4007-8a71-01aa312ed467")
<BulkTouchQuerySet [<Package (pulp_type=rpm.package): pk=0000057e-3b84-4007-8a71-01aa312ed467>]>
>>> list(v.repository.latest_version().get_content2(Package.objects).order_by("-pk").values_list("pk")) == list(v.repository.latest_version().get_content(Package.objects).order_by("-pk").values_list("pk"))
True
>>> v.repository.latest_version()
<Repository: @rubygems/rubygems/fedora-rawhide-x86_64; Version: 3>

```

File: CHANGES/3969.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Attained the two-fold speedup for the ``RepositoryVersion.get_content()`` filter.
```
Comment (gerrod3) [inline CHANGES/3969.bugfix:1]:
```suggestion
Attained a two-fold speedup for the ``RepositoryVersion.get_content()`` filter.
```

**Review by gerrod3 — APPROVED** [review]

**Comment by dralley** [comment]
@lubosmj Well my idea would be to have a script that would just go through each of the repositories, and each of the versions of that repository, and compare rather than spot-checking a couple.  And logging the time difference in a way that we could plot it later.

But otherwise yes basically doing what you did, just automated and against more data.  

We could also check couple of repos that have had versions deleted at the beginning, middle, and/or end, but there's no need for that part to be automated.

**Comment by lubosmj** [comment]
I have measured and compared the outputs across 3090 repositories (with at least 100 content units. All the outputs were equal between the old and updated versions of `get_content`. The repositories were synced on demand.

image

The red dots signify the the old version of `get_content`. The blue ones represent the new version of `get_content`.

Measured code:
```
round(timeit.timeit("v.get_content(Package.objects).count()", setup="from pulp_rpm.app.models import RpmRepository, Package", globals=locals(), number=10), 3)
```

Test script:
```
import timeit

from pulp_rpm.app.models import RpmRepository, Package


def test():
    print("{:<75} {:<6} {:<6} {:<6} {:<5} {:<5}".format("Repository (version)", "Units", "Content", "Count", "Time (or)", "Time (up)"))
    for repo in RpmRepository.objects.all():
        for v in repo.versions.all():
            packages = v.get_content(Package.objects).count()
            if packages > 100:
                # Comparing the content in version {v.number}...
                equals = list(v.get_content2(Package.objects).order_by("pk").values_list("pk")) == list(v.get_content(Package.objects).order_by("pk").values_list("pk"))
                content = "OK" if equals else "ERROR"
                # Comparing the number of content units ({packages})...
                equals = v.get_content2(Package.objects).count() == packages
                count = "OK" if equals else "ERROR"
                time_original = round(timeit.timeit("v.get_content(Package.objects).count()", setup="from pulp_rpm.app.models import RpmRepository, Package", globals=locals(), number=10), 3)
                time_updated = round(timeit.timeit("v.get_content2(Package.objects).count()", setup="from pulp_rpm.app.models import RpmRepository, Package", globals=locals(), number=10), 3)
                print("{:<75} {:<6} {:<6} {:<6} {:<5} {:<5}".format(repo.name + f" ({v.number})", packages, content, count, time_original, time_updated))


test()
```

**Comment by dralley** [comment]
Heck yeah.  And that's exactly why I wanted the graphs :). That's a solid, solid improvement

File: CHANGES/3969.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Attained a two-fold speedup for the ``RepositoryVersion.get_content()`` filter.
```
[the remark this answers is not in the document]
Comment (daviddavis) [reply CHANGES/3969.bugfix:1]: On the other hand, maybe this change is for plugin writers? In which case maybe it belongs in the plugin changelog.

[the remark this answers is not in the document]
Comment (dralley) [reply CHANGES/3969.bugfix:1]: I agree that the user-facing changelog should be more general, 

> Improved performance throughout the app when looking up content for repository versions

would be good enough for me.

**Review by ipanova — APPROVED** [review]

**Comment by dralley** [comment]
Copying from the issue:

Before:

```sql
(0.198) EXPLAIN (ANALYZE true) SELECT "rpm_package"."content_ptr_id"
FROM "rpm_package"
INNER JOIN "core_content" ON ("rpm_package"."content_ptr_id" = "core_content"."pulp_id")
INNER JOIN "core_repositorycontent" ON ("core_content"."pulp_id" = "core_repositorycontent"."content_id")
WHERE "core_repositorycontent"."pulp_id" IN (
  SELECT U0."pulp_id"
  FROM "core_repositorycontent" U0
  INNER JOIN "core_repositoryversion" U2 ON (U0."version_added_id" = U2."pulp_id")
  LEFT OUTER JOIN "core_repositoryversion" U3 ON (U0."version_removed_id" = U3."pulp_id")
  WHERE (U0."repository_id" = '0189b6fd82f878a1a9a0b3dab55e8684'::uuid AND U2."number" <= 1 AND NOT (U3."number" <= 1 AND U3."number" IS NOT NULL))
);
args=(UUID('0189b6fd-82f8-78a1-a9a0-b3dab55e8684'), Int4(1), Int4(1));
alias=default
```

After:

```
(0.073) EXPLAIN (ANALYZE true) SELECT "rpm_package"."content_ptr_id"
FROM "rpm_package"
WHERE "rpm_package"."content_ptr_id" IN (
  SELECT U0."content_id"
  FROM "core_repositorycontent" U0
  INNER JOIN "core_repositoryversion" U2 ON (U0."version_added_id" = U2."pulp_id")
  LEFT OUTER JOIN "core_repositoryversion" U3 ON (U0."version_removed_id" = U3."pulp_id")
  WHERE (U0."repository_id" = '0189b6fd82f878a1a9a0b3dab55e8684'::uuid AND U2."number" <= 1 AND NOT (U3."number" <= 1 AND U3."number" IS NOT NULL))
);
args=(UUID('0189b6fd-82f8-78a1-a9a0-b3dab55e8684'), Int4(1), Int4(1));
alias=default
```

**Pull request merged**